File: app/lib/models/verse.dart
import 'json.dart';

/// Who rendered a translation. Only ever shown next to their own text — an
/// acharya's purport and our own explanation are different things and the UI
/// must never let them look alike.
class Translator {
  const Translator({required this.id, required this.slug, required this.name, this.imageUrl});

  final String id;
  final String slug;
  final String name;
  final String? imageUrl;

  factory Translator.fromJson(Json json) => Translator(
        id: asString(json['id']),
        slug: asString(json['slug']),
        name: asString(json['name']),
        // Reference rows send `imageUrl`; the nested translator sends the raw
        // `imagePath`, which is not fetchable — so only a signed URL is kept.
        imageUrl: asStringOrNull(json['imageUrl']),
      );
}

/// One rendering of a verse, in one language, by one translator.
class VerseTranslation {
  const VerseTranslation({
    required this.id,
    required this.languageCode,
    required this.type,
    this.meaning,
    this.purport,
    this.sourceRef,
    this.translator,
  });

  final String id;
  final String languageCode;
  final String type;
  final String? meaning;

  /// The commentary. Long — screens should treat it as an expandable section,
  /// not something to render inside a card.
  final String? purport;
  final String? sourceRef;
  final Translator? translator;

  bool get hasPurport => (purport ?? '').trim().isNotEmpty;

  factory VerseTranslation.fromJson(Json json) => VerseTranslation(
        id: asString(json['id']),
        languageCode: asString(json['languageCode']),
        type: asString(json['type']),
        meaning: asStringOrNull(json['meaning']),
        purport: asStringOrNull(json['purport']),
        sourceRef: asStringOrNull(json['sourceRef']),
        translator: asJson(json['translator']) == null
            ? null
            : Translator.fromJson(asJson(json['translator'])!),
      );
}

/// Our own plain-language note. Deliberately a different model from
/// [VerseTranslation] so it can never be presented as commentary.
class VerseExplanation {
  const VerseExplanation({required this.text, required this.source});

  final String text;

  /// `EDITORIAL` or `AI` — the UI labels the second one as such.
  final String source;

  bool get isGenerated => source.toUpperCase() == 'AI';

  factory VerseExplanation.fromJson(Json json) => VerseExplanation(
        text: asString(json['text']),
        source: asString(json['source'], 'EDITORIAL'),
      );
}

/// The book and chapter a verse sits in, as sent inline with it.
class VerseBookRef {
  const VerseBookRef({required this.id, required this.slug, required this.title, required this.bookNumber});

  final String id;
  final String slug;
  final String title;
  final int bookNumber;

  factory VerseBookRef.fromJson(Json json) => VerseBookRef(
        id: asString(json['id']),
        slug: asString(json['slug']),
        title: asString(json['title']),
        bookNumber: asInt(json['bookNumber']),
      );
}

class VerseChapterRef {
  const VerseChapterRef({required this.id, required this.number, required this.title});

  final String id;
  final int number;
  final String title;

  factory VerseChapterRef.fromJson(Json json) => VerseChapterRef(
        id: asString(json['id']),
        number: asInt(json['number']),
        title: asString(json['title']),
      );
}

/// A verse, already resolved into the reader's languages by the API.
class Verse {
  const Verse({
    required this.id,
    required this.verseId,
    required this.bookNumber,
    required this.type,
    this.cantoNumber,
    this.chapterNumber,
    this.verseNumber,
    this.verseNumberEnd,
    this.sanskrit,
    this.transliteration,
    this.wordMeanings,
    this.audioUrl,
    this.tags = const [],
    this.book,
    this.chapter,
    this.translation,
    this.availableTranslations = const [],
    this.explanation,
  });

  final String id;

  /// The human reference — `1.2.13` for Gita, `2.10.1.5-7` for a Bhagavatam
  /// range. Every scraped file is keyed on it, so it is what the app routes on.
  final String verseId;
  final int bookNumber;
  final String type;
  final int? cantoNumber;
  final int? chapterNumber;
  final int? verseNumber;

  /// Set when the verse covers a range, e.g. 5–7 presented as one.
  final int? verseNumberEnd;

  final String? sanskrit;
  final String? transliteration;
  final String? wordMeanings;
  final String? audioUrl;
  final List<String> tags;

  final VerseBookRef? book;
  final VerseChapterRef? chapter;

  /// The rendering chosen for this reader.
  final VerseTranslation? translation;

  /// The others they can switch to, without their text.
  final List<VerseTranslation> availableTranslations;

  final VerseExplanation? explanation;

  /// "Bhagavad Gita · 2.13" — the citation, with the book set off from the
  /// numbering. The same parts as [reference], punctuated for display.
  String get citation {
    final title = book?.title;
    final numbering = _numbering;
    if (title == null || title.isEmpty) return numbering.isEmpty ? verseId : numbering;
    return numbering.isEmpty ? title : '$title \u00B7 $numbering';
  }

  /// "Bhagavad Gita 2.13" — what a card shows under the Sanskrit.
  String get reference {
    final title = book?.title;
    final numbering = _numbering;
    if (title == null || title.isEmpty) return numbering.isEmpty ? verseId : numbering;
    return numbering.isEmpty ? title : '$title $numbering';
  }

  /// "2.13", or "2.10.1.5-7" for a Bhagavatam range.
  String get _numbering => <String>[
        if (cantoNumber != null) '$cantoNumber',
        if (chapterNumber != null) '$chapterNumber',
        if (verseNumber != null)
          verseNumberEnd != null ? '$verseNumber-$verseNumberEnd' : '$verseNumber',
      ].join('.');

  factory Verse.fromJson(Json json) => Verse(
        id: asString(json['id']),
        verseId: asString(json['verseId']),
        bookNumber: asInt(json['bookNumber']),
        type: asString(json['type']),
        cantoNumber: asIntOrNull(json['cantoNumber']),
        chapterNumber: asIntOrNull(json['chapterNumber']),
        verseNumber: asIntOrNull(json['verseNumber']),
        verseNumberEnd: asIntOrNull(json['verseNumberEnd']),
        sanskrit: asStringOrNull(json['sanskrit']),
        transliteration: asStringOrNull(json['transliteration']),
        wordMeanings: asStringOrNull(json['wordMeanings']),
        audioUrl: asStringOrNull(json['audioUrl']),
        tags: asStringList(json['tags']),
        book: asJson(json['book']) == null ? null : VerseBookRef.fromJson(asJson(json['book'])!),
        chapter: asJson(json['chapter']) == null
            ? null
            : VerseChapterRef.fromJson(asJson(json['chapter'])!),
        translation: asJson(json['translation']) == null
            ? null
            : VerseTranslation.fromJson(asJson(json['translation'])!),
        availableTranslations: asList(json['availableTranslations'], VerseTranslation.fromJson),
        explanation: asJson(json['explanation']) == null
            ? null
            : VerseExplanation.fromJson(asJson(json['explanation'])!),
      );
}

File: backend/routes/admin/verse.js
import { createRouter, z, schemas } from '../../utils/router.js';
import * as controller from '../../controllers/admin/verse.js';

const router = createRouter({
  tag: 'Admin · Verses',
  prefix: '/verses',
  description:
    'Verses and everything attached to them. Two rules run through all of it: `verseId` is ' +
    'immutable, because deeplinks and bookmarks already point at it; and translations ' +
    '(an acharya’s work) are never stored alongside explanations (ours, or the AI’s).',
});

const verseIdParam = z.object({
  verseId: z.string().regex(/^\d+(\.\d+){1,3}(-\d+)?$/, 'Expected a dotted verse id'),
});

router.get(
  '/',
  {
    summary: 'List verses',
    description:
      'The editorial worklist. `untranslated=true` returns verses with no published rendering ' +
      'at all, which is usually what someone opening this screen is looking for.',
    permission: 'verse.read',
    limit: 'read',
    query: schemas.page.extend({
      bookId: z.string().optional(),
      canto: z.coerce.number().int().optional(),
      chapter: z.coerce.number().int().optional(),
      isSlokaEligible: z.coerce.boolean().optional(),
      untranslated: z.coerce.boolean().optional(),
    }),
    responds: { 200: 'A page of verses with attachment counts' },
  },
  controller.list
);

router.get(
  '/:verseId',
  {
    summary: 'Get one verse with everything attached',
    description: 'Including unpublished translations, explanations, narrations and issue mappings.',
    permission: 'verse.read',
    limit: 'read',
    params: verseIdParam,
    responds: { 200: 'The verse', 404: 'No such verse' },
  },
  controller.get
);

router.post(
  '/',
  {
    summary: 'Create a verse',
    permission: 'verse.write',
    limit: 'write',
    body: z.object({
      verseId: z.string().regex(/^\d+(\.\d+){1,3}(-\d+)?$/),
      bookId: z.string().min(1),
      chapterId: z.string().nullable().optional(),
      cantoNumber: z.coerce.number().int().nullable().optional(),
      chapterNumber: z.coerce.number().int().nullable().optional(),
      verseNumber: z.coerce.number().int().min(1),
      verseNumberEnd: z.coerce.number().int().nullable().optional(),
      type: z.enum(['SHLOKA', 'PROSE', 'LINE', 'REFRAIN', 'MANTRA']).optional(),
      sanskrit: z.string().optional(),
      transliteration: z.string().optional(),
      wordMeanings: z.any().optional(),
      audioPath: z.string().max(500).nullable().optional(),
      tags: z.array(z.string().max(50)).optional(),
    }),
    responds: { 201: 'The verse', 404: 'No such book' },
  },
  controller.create
);

router.patch(
  '/:verseId',
  {
    summary: 'Update a verse',
    description: '`verseId` itself cannot be changed — links that already exist point at it.',
    permission: 'verse.write',
    limit: 'write',
    params: verseIdParam,
    body: z.object({
      sanskrit: z.string().optional(),
      transliteration: z.string().optional(),
      wordMeanings: z.any().optional(),
      audioPath: z.string().max(500).nullable().optional(),
      tags: z.array(z.string().max(50)).optional(),
      type: z.enum(['SHLOKA', 'PROSE', 'LINE', 'REFRAIN', 'MANTRA']).optional(),
      isSlokaEligible: z.boolean().optional(),
    }),
    responds: { 200: 'The verse', 400: 'Attempted to change verseId', 404: 'No such verse' },
  },
  controller.update
);

router.delete(
  '/:verseId',
  {
    summary: 'Delete a verse',
    permission: 'verse.delete',
    limit: 'write',
    params: verseIdParam,
    responds: { 204: 'Deleted', 404: 'No such verse' },
  },
  controller.remove
);

router.post(
  '/sloka-eligibility',
  {
    summary: 'Mark verses as sloka-eligible, in bulk',
    description:
      'Curating this pool is done a chapter at a time, so it takes a list. Only Bhagavad Gita ' +
      'and Srimad Bhagavatam verses are ever eligible — a short work is not something to hand ' +
      'someone as the verse for their day, and the restriction is enforced here rather than ' +
      'trusted to whoever is clicking.',
    permission: 'sloka.manage',
    limit: 'write',
    body: z.object({
      verseIds: z.array(z.string()).min(1).max(500),
      isSlokaEligible: z.boolean(),
    }),
    responds: { 200: 'How many changed', 400: 'A verse from an ineligible book', 404: 'Unknown verse ids' },
  },
  controller.setSlokaEligibility
);

router.put(
  '/:verseId/translations',
  {
    summary: 'Add or replace a translation',
    description:
      'One translator’s rendering in one language. Keyed on verse + translator + language + ' +
      'type, so re-sending the same combination edits it rather than duplicating it.',
    permission: 'verse.write',
    limit: 'write',
    params: verseIdParam,
    body: z.object({
      translatorId: z.string().min(1),
      languageCode: z.string().max(10),
      type: z.enum(['TRANSLATION', 'COMMENTARY', 'POETIC_EXPANSION']).optional(),
      meaning: z.string().min(1),
      purport: z.string().optional(),
      sourceRef: z.string().max(50).optional(),
      audioPath: z.string().max(500).nullable().optional(),
      isPublished: z.boolean().optional(),
      displayOrder: z.coerce.number().int().optional(),
    }),
    responds: { 200: 'The translation', 404: 'No such verse' },
  },
  controller.upsertTranslation
);

router.delete(
  '/:verseId/translations/:translationId',
  {
    summary: 'Delete a translation',
    permission: 'verse.delete',
    limit: 'write',
    params: z.object({ verseId: z.string(), translationId: z.string() }),
    responds: { 204: 'Deleted' },
  },
  controller.deleteTranslation
);

router.put(
  '/:verseId/explanation',
  {
    summary: 'Add or replace an explanation',
    description:
      'A short plain-language note, written once per language and reused by every reader. ' +
      'Deliberately not stored as a translation: app-written or generated text must never be ' +
      'able to surface as though an acharya had said it.',
    permission: 'verse.write',
    limit: 'write',
    params: verseIdParam,
    body: z.object({
      languageCode: z.string().max(10),
      text: z.string().min(1).max(5000),
      source: z.enum(['MANUAL', 'RULE', 'AI']).optional(),
      isPublished: z.boolean().optional(),
    }),
    responds: { 200: 'The explanation', 404: 'No such verse' },
  },
  controller.upsertExplanation
);

router.post(
  '/:verseId/narrations',
  {
    summary: 'Add a narration',
    description:
      'A saint’s telling of the story around a verse, attributed and usually with audio. ' +
      'Narrative, not commentary on the text.',
    permission: 'verse.write',
    limit: 'write',
    params: verseIdParam,
    body: z.object({
      guruId: z.string().nullable().optional(),
      languageCode: z.string().max(10),
      title: z.string().max(200).optional(),
      text: z.string().optional(),
      audioPath: z.string().max(500).nullable().optional(),
      durationMs: z.coerce.number().int().optional(),
      sourceNote: z.string().max(500).optional(),
      isPublished: z.boolean().optional(),
    }),
    responds: { 201: 'The narration', 404: 'No such verse' },
  },
  controller.createNarration
);

router.patch(
  '/:verseId/narrations/:narrationId',
  {
    summary: 'Update a narration',
    permission: 'verse.write',
    limit: 'write',
    params: z.object({ verseId: z.string(), narrationId: z.string() }),
    body: z.object({
      title: z.string().max(200).optional(),
      text: z.string().optional(),
      audioPath: z.string().max(500).nullable().optional(),
      durationMs: z.coerce.number().int().optional(),
      isPublished: z.boolean().optional(),
      displayOrder: z.coerce.number().int().optional(),
    }),
    responds: { 200: 'The narration' },
  },
  controller.updateNarration
);

router.delete(
  '/:verseId/narrations/:narrationId',
  {
    summary: 'Delete a narration',
    permission: 'verse.delete',
    limit: 'write',
    params: z.object({ verseId: z.string(), narrationId: z.string() }),
    responds: { 204: 'Deleted' },
  },
  controller.deleteNarration
);

router.post(
  '/:verseId/links',
  {
    summary: 'Link one verse to another',
    description:
      'Directional. "Expands on" does not read the same in reverse, so A→B and B→A are ' +
      'separate rows and the editor decides whether both should exist.',
    permission: 'verse.write',
    limit: 'write',
    params: verseIdParam,
    body: z.object({
      targetVerseId: z.string().min(1),
      relation: z.enum(['SAME_CONCEPT', 'EXPANDS_ON', 'QUOTED_IN', 'CONTRASTS_WITH']),
      note: z.string().max(500).optional(),
    }),
    responds: { 201: 'The link', 400: 'A verse cannot link to itself', 404: 'No such verse' },
  },
  controller.createLink
);

router.delete(
  '/:verseId/links/:linkId',
  {
    summary: 'Remove a link',
    permission: 'verse.write',
    limit: 'write',
    params: z.object({ verseId: z.string(), linkId: z.string() }),
    responds: { 204: 'Deleted' },
  },
  controller.deleteLink
);

router.put(
  '/:verseId/issues',
  {
    summary: 'Set which struggles a verse speaks to',
    description:
      'The table the personalised sloka reads. `weight` decides which verse wins when several ' +
      'speak to the same struggle: 90+ means the verse addresses it head on, 50 means it is ' +
      'relevant. Replaces the whole set for this verse. This is the highest-leverage ' +
      'editorial surface in the system — a handful of well-chosen verses per struggle does ' +
      'more than a complete but careless mapping.',
    permission: 'sloka.manage',
    limit: 'write',
    params: verseIdParam,
    body: z.object({
      issues: z
        .array(
          z.object({
            slug: z.string().min(1).max(100),
            weight: z.coerce.number().int().min(0).max(100).optional(),
          })
        )
        .max(20),
    }),
    responds: { 200: 'The mapping', 400: 'Unknown issue slug', 404: 'No such verse' },
  },
  controller.setIssues
);

export default router;
